### DISCUSSION TITLE
Precompilation for production use cases

### DISCUSSION THREAD

**Discussion by Daniel15** [pyscript/pyscript#1168] [Proposals]
For production use cases, it'd be useful to:
 - Have the ability to compile the app's Python code to WebAssembly as part of a build process.
 - Implement some form of tree shaking such that the unused parts of the Python libraries are removed as part of the build process, leaving just the code that's actually needed.

This could probably make it viable for some real use cases rather than just as an experiment 😄

**Comment by verhulstm** [comment]
hi,

yes we already can compile the app's Python code to WebAssembly

check out

https://github.com/pyscript/pyscript/issues/85

**Comment by Daniel15** [comment]
Sounds good. That just leaves tree shaking I guess, similar to what Blazor does (since Microsoft don't want Blazor apps to ship the entire .NET Framework to the user, as it's quite large, it removes bits that aren't used as part of the build)

**Comment by verhulstm** [comment]
yup, work is being done on tree shaking

check this out

https://jott.live/markdown/wasm_binary_size 

and we are looking into other things to do to

**Comment by fpliger** [comment]
Great suggestions, thanks @Daniel15 !

> Have the ability to compile the app's Python code to WebAssembly as part of a build process.

I think the problem with this is that the problem is not only pre-compiling but also running it for the first time. It takes time. (I may be wrong here though). 

> Implement some form of tree shaking such that the unused parts of the Python libraries are removed as part of the build process, leaving just the code that's actually needed.

There's work being done about that upstream. For instance [pyodide-pack](https://github.com/rth/pyodide-pack) is a great example. 

> This could probably make it viable for some real use cases rather than just as an experiment 😄

This is less about "real use cases" and more about being viable to specific use cases. PyScript is much of an experiment as any project so early in its development. There are specific use cases that will never be good ones for PyScript but there are many use cases where PyScript is a viable option today. 

If you can elaborate on your use cases it could definitely be interesting for the core team to take into consideration as PyScript matures.
